builtins/web/fetch/fetch-api.cpp

```diff
@@ -96,7 +99,273 @@ bool fetch(JSContext *cx, unsigned argc, Value *vp) {
+    RootedValue body_val(cx, JS::ObjectValue(*blob));
+    RootedValue init_val(cx);
+    if (!Response::initialize(cx, response_obj, body_val, init_val)) {
+      return false;
+    }
+
+    // 2. Set response's status message to `OK`.
+    Response::set_status_message_from_code(cx, response_obj, 200);
+
+    // 3. Set response's body to bodyWithType's body.
+    // 4. Set response's header list to (`Content-Length`, serializedFullLength), (`Content-Type`, type).
+    // 3 and 4 done at the end.
+  // 14. Otherwise:
+  } else {
+    // 1. Set response's range-requested flag.
+    // 2. Let rangeHeader be the result of getting `Range` from request's header list.
+    // 3. Let rangeValue be the result of parsing a single range header value given rangeHeader and true.
+    // 4. If rangeValue is failure, then return a network error.
+    auto *headers = Headers::get_index(cx, req_headers, maybe_range_index.value());
+    MOZ_ASSERT(headers);
+
+    auto &[key, val] = *headers;
+    std::string_view range_value(val);
+
+    if (!range_value.starts_with("bytes=")) {
+      return api::throw_error(cx, FetchErrors::FetchNetworkError);
+    }
+
+    range_value.remove_prefix(6); // bytes=
+    auto dash_pos = range_value.find('-');
+
+    if (dash_pos == std::string_view::npos) {
+      return api::throw_error(cx, FetchErrors::FetchNetworkError);
+    }
+
+    auto start_str = range_value.substr(0, dash_pos);
+    auto end_str = range_value.substr(dash_pos + 1);
+
+    auto to_size = [](std::string_view s) -> std::optional<size_t> {
+      size_t v;
+      auto [ptr, ec] = std::from_chars(s.begin(), s.end(), v);
+      return ec == std::errc() ? std::optional<size_t>(v) : std::nullopt;
+    };
+
+    //   5. Let (rangeStart, rangeEnd) be rangeValue.
+    auto maybe_start_range = to_size(start_str);
+    auto maybe_end_range = to_size(end_str);
+
+    size_t start_range = 0;
+    size_t end_range = 0;
+
+    // 6. If rangeStart is null:
+    if (!maybe_start_range.has_value()) {
+      // If both start_range and end_range are not provided, it's an error.
+      if (!maybe_end_range.has_value()) {
+        return api::throw_error(cx, FetchErrors::FetchNetworkError);
+      }
+
+      // 1. Set rangeStart to fullLength - rangeEnd.
+      // 2. Set rangeEnd to rangeStart + rangeEnd - 1.
+      start_range = full_len - maybe_end_range.value();
+      end_range = start_range + maybe_end_range.value() - 1;
+      // 7. Otherwise:
+    } else {
+      // 1. If rangeStart is greater than or equal to fullLength, then return a network error.
+      if (maybe_start_range.value() > full_len) {
+        return api::throw_error(cx, FetchErrors::FetchNetworkError);
+      }
+      // 2. If rangeEnd is null or rangeEnd is greater than or equal to fullLength, then set
+      // rangeEnd to fullLength - 1.
+      start_range = maybe_start_range.value();
+      end_range = std::min(maybe_end_range.value_or(full_len - 1), full_len - 1);
+    }
```

Comment:
Could we refactor the above into a separate public function `std::tuple<size_t, size_t> Blob::extract_range(blob, std::string_view range_query)` or similar?

Reply:
I think this could be done for any resource not blob only. I moved this to `fetch-utils.h` file. I hope it's fine :)

Reply:
Thanks, yes that works and will help us a lot to implement this at Fastly as we build our own request-response.cpp file implementation with Fastly-specific handling.